### PR TITLE
[WIP] enable QLoRA + FSDP2

### PR DESCRIPTION
**Pull request by weifengpy**
this PR is stacked on LoRA + FSDP2: https://github.com/pytorch/torchtune/pull/855

they share the components
* loading HF checkpoint
* FSDP wrapping
* saving checkpoint
* load checkpoint to resume training

the major difference is in config
* `model._component_`: `torchtune.models.llama2.qlora_llama2_7b` instead of `lora_llama2_7b`. `LoRALinear(quantize_base=True/False)`

### CODE REVIEW

File: tests/torchtune/utils/test_distributed.py [hunk 1]
```diff
@@ -245,3 +251,160 @@ def test_lora_meta_device_init_fsdp(self):
...
+
+        # inp is different for each rank
+        torch.manual_seed(42 + self.rank)
+
+        # test get full state dict
+        for _ in range(epochs):
+            inp = torch.randn((2, mlp_dim), device="cuda")
+            base_model(inp).sum().backward()
+            for param in base_model.parameters():
+                torch.distributed.all_reduce(param.grad)
```
Comment (awgu) [inline tests/torchtune/utils/test_distributed.py:294]: nit: 😄 
```suggestion
                torch.distributed.all_reduce(param.grad, op=ReduceOp.AVG)
```

`from torch.distributed.distributed_c10d import ReduceOp`

Comment (weifengpy) [reply tests/torchtune/utils/test_distributed.py:294]: neat!

Comment (weifengpy) [reply tests/torchtune/utils/test_distributed.py:294]: updated LoRA PR to use `ReduceOp.AVG`. This PR will be stacked on it when landing

File: torchtune/modules/low_precision/_register_nf4_dispatch_ops.py [hunk 2]
```diff
@@ -17,26 +17,3 @@ def clone(func, *args, **kwargs):
     in precision.
     """
     return to_nf4(args[0][0].get_original_weight())
-
-
```
Comment (weifengpy) [inline torchtune/modules/low_precision/_register_nf4_dispatch_ops.py:21]: starting from TorchAO==0.2.0, we implemented `NF.copy_`. It's the superset of `inplace_copy`. We cover both `NF4.copy_(bf16)` and `NF4.copy_(NF4)`

File: recipes/configs/dev/llama2/7B_qlora.yaml [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+  lr: 3e-4
+lr_scheduler:
+  _component_: torchtune.modules.get_cosine_schedule_with_warmup
+  num_warmup_steps: 100
+
+loss:
+  _component_: torch.nn.CrossEntropyLoss
+
+fsdp:
+  cpu_offload: False
```
Comment (weifengpy) [inline recipes/configs/dev/llama2/7B_qlora.yaml:65]: comparing with `7B_qlora_single_device.yaml`, `cpu_offload` is the new feature

File: recipes/configs/dev/llama2/7B_qlora.yaml [hunk 4]
```diff
@@ -0,0 +1,91 @@
+# Config for single device QLoRA with lora_finetune_single_device.py
+# using a Llama2 7B model
+#
+# This config assumes that you've run the following command before launching
+# this run:
+#   tune download meta-llama/Llama-2-7b-hf --output-dir /tmp/Llama-2-7b-hf --hf-token <HF_TOKEN>
+#
+# To launch on a single device, run the following command from root:
+#   tune run lora_finetune_single_device --config llama2/7B_qlora_single_device
```
Comment (weifengpy) [inline recipes/configs/dev/llama2/7B_qlora.yaml:9]: will replace `lora_finetune_single_device` to `lora_finetune_fsdp2` before landing. saving CI to publish PR earlier now

File: recipes/dev/lora_finetune_fsdp2.py [hunk 5]
```diff
@@ -290,25 +291,6 @@ def _setup_model(
         with utils.set_default_dtype(self._dtype), torch.device("meta"):
             model = config.instantiate(cfg_model)
 
-        if self._is_rank_zero:
-            # The model contains LoRA params which won't have any matching keys in
-            # the state dict. As a result, we need to load with strict=False.
-            # Before loading the state dict, ensure the state dict keys for the base
-            # model and adapters (if available) match the keys in the full LoRA model
-            # This is a good sanity check to prevent silent errors
-            validate_state_dict_for_lora(
```
Comment (weifengpy) [inline recipes/dev/lora_finetune_fsdp2.py:299]: `validate_state_dict_for_lora` needs some refactor since FSDP2 use clean FQNs. 
FSDP2 also check param on meta and throw Runtime error https://fburl.com/wrhxykn3

```
"FSDP parameters should be materialized from meta device before training, "
f"but the following were still on meta device: {param_names_on_meta}\n"
"For example, call module.to_empty(device) to materialize to device and "
"call module.reset_parameters() on each module to initialize values."
```

File: recipes/dev/lora_finetune_fsdp2.py [hunk 6]
```diff
@@ -590,6 +577,8 @@ def train(self) -> None:
                 logits = logits.transpose(1, 2)
                 # Compute loss
                 loss = self._loss_fn(logits, labels)
+                # free logits otherwise it peaks backward memory
+                del logits
```
Comment (weifengpy) [inline recipes/dev/lora_finetune_fsdp2.py:581]: `del logits` saves 6% peak memory (baseline has `cpu_offload=True`). see https://github.com/pytorch/torchtune/pull/1046

File: recipes/configs/dev/llama2/7B_qlora.yaml (hunk 3 above)
Comment (ebsmothers) [reply recipes/configs/dev/llama2/7B_qlora.yaml:65]: Strictly speaking you probably do not need to put `cpu_offload` config under `fsdp` (unless you think it helps for clarity, or anticipate other kwargs that we'd wanna pass through to `fully_shard`). No strong preference either way here though

File: torchtune/utils/_distributed.py [hunk 7]
```diff
@@ -280,12 +281,39 @@ def load_from_full_model_state_dict(
...
+                sharded_meta_param.device_mesh,
+                sharded_meta_param.placements,
+                shape=sharded_meta_param.size(),
+                dtype=sharded_meta_param.dtype,
+                requires_grad=sharded_meta_param.requires_grad,
+                stride=sharded_meta_param.stride(),
+            )
+        else:
+            sharded_tensor = distribute_tensor(
+                full_tensor,
```
Comment (ebsmothers) [inline torchtune/utils/_distributed.py:313]: No longer need to convert dtype as before?

File: recipes/configs/dev/llama2/70B_qlora.yaml [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+# This config assumes that you've run the following command before launching
+# this run:
+#   tune download meta-llama/Llama-2-70b-hf --output-dir /tmp/Llama-2-70b-hf --hf-token <HF_TOKEN>
+#
+# This config needs 8 GPUs to run
+#   # tune run --nproc_per_node 8 lora_finetune_fsdp2 --config llama2/70B_qlora
+#
+
+# Model Arguments
+model:
```
Comment (weifengpy) [inline recipes/configs/dev/llama2/70B_qlora.yaml:13]: comparing with `70B_lora.yaml`, added `output_proj`, `apply_lora_to_mlp: True`

File: torchtune/utils/_distributed.py [hunk 9]
```diff
@@ -280,12 +281,39 @@ def load_from_full_model_state_dict(
...
+            sharded_param[: chunk.size(0)].copy_(chunk)
+            sharded_tensor = DTensor(
+                sharded_param,
+                sharded_meta_param.device_mesh,
+                sharded_meta_param.placements,
+                shape=sharded_meta_param.size(),
+                dtype=sharded_meta_param.dtype,
+                requires_grad=sharded_meta_param.requires_grad,
+                stride=sharded_meta_param.stride(),
+            )
```
Comment (ebsmothers) [inline torchtune/utils/_distributed.py:310]: If it's not too much work, is it possible to extend the unit test you added in #855 to cover this case as well? As is I find it a bit hard to follow and want to make sure we have a reliable sanity check in case anything breaks in the future

Comment (weifengpy) [reply torchtune/utils/_distributed.py:310]: you're right. will try to come up with a unit test to cover NF4

File: torchtune/utils/_distributed.py (hunk 7 above)
Comment (weifengpy) [reply torchtune/utils/_distributed.py:313]: it's moved to line 284 `full_tensor = full_tensor.to(sharded_meta_param.dtype).to(device)`. The extra `.to(device)` is for NF4 since quantization on cpu is prohibitively slow

File: recipes/configs/dev/llama2/7B_qlora.yaml (hunk 3 above)
Comment (weifengpy) [reply recipes/configs/dev/llama2/7B_qlora.yaml:65]: I know Less has been working on `cpu_offload` for activations, which might be a root config

another things is, other `fully_shard` kwargs like `reshard_after_forward=False` can boost QPS

Does it make sense to you?

File: recipes/configs/dev/llama2/7B_qlora.yaml [hunk 10]
```diff
@@ -0,0 +1,91 @@
+# Config for single device QLoRA with lora_finetune_single_device.py
```
[the remark this answers is not in the document]
Comment (weifengpy) [reply recipes/configs/dev/llama2/7B_qlora.yaml]: good point! updating now

File: torchtune/utils/_distributed.py (hunk 9 above)
Comment (weifengpy) [reply torchtune/utils/_distributed.py:310]: @ebsmothers see `pytest -s tests/torchtune/utils/test_distributed.py -k test_qlora_state_dict`

File: recipes/dev/lora_finetune_fsdp2.py (hunk 5 above)
Comment (weifengpy) [reply recipes/dev/lora_finetune_fsdp2.py:299]: `validate_missing_and_unexpected_for_lora` works out of the box. added it in this PR. finger acrossed!

File: recipes/configs/dev/llama2/7B_qlora.yaml (hunk 3 above)
Comment (ebsmothers) [reply recipes/configs/dev/llama2/7B_qlora.yaml]: Ah yeah this is a good point, makes sense to me. Thanks for clarifying

**Review by ebsmothers — APPROVED** [review]
This is looking good! As discussed offline we can separate out the checkpoint save support into a follow-up PR to unblock other ongoing work. Once CI is green this one should be good to land. Thank you for enabling a huge huge feature for our users!

**Comment by weifengpy** [comment]
merged after confirming CI's all green. Will follow up on saving state dict
